test/socketio.js

```diff
@@ -0,0 +1,129 @@
+"use strict";
+
+const request = require("request-promise");
+const tough = require("tough-cookie");
+const users = require("./data/users");
+const classrooms = require("./data/classrooms");
+
+const io = require("socket.io-client");
+const baseUrl = 'http://localhost:3000';
+const options = {  
+  transports: ['websocket'],
+  'force new connection': true
+};
+
+let authToken;
+const sleep = (t) => new Promise((r) => setTimeout(r, t));
+
+// This series of tests uses user[2] for registration. login, and manipulation
+describe("Socket.IO", () => {
+    let client1, client2, client3;
+
+    it("should connect to the server successfully", (done) => {
+        client1 = io(baseUrl, options);
+        client1.on("connect",()=>{
+            expect(client1.connected).to.equal(true);
+            client1.disconnect();
+            done();
+        });
+        sleep(2000)
+        .then(_=>{client1.disconnect(); done(new Error("Timed out"))});
+    });
+
+    it("should login to server using token", (done) => {
+        const thisUser = users[3];
+        const j = request.jar();
+        const loginOpts = {
+            method: "POST",
+            uri: baseUrl + "/login",
+            json: true,
+            body: {
+                email: thisUser.email,
+                password: thisUser.password,
+            },
+            jar: j,
+        };
+        const authOpts = {
+            method: "POST",
+            uri: baseUrl + "/api/v1/authenticate",
+            json: true,
+            jar: j,
+            resolveWithFullResponse: true,
+        };
+
+        request(loginOpts)
+        .then( (user) => {
+            expect(user.firstName).to.equal(thisUser.firstName);
+            expect(user.lastName).to.equal(thisUser.lastName);
+            return request(authOpts);
+        }).then( (auth) => {
+            expect(auth.statusCode).to.equal(200);
+            authToken = auth.body.token;
+            client2 = io(baseUrl, options);
+            client2.on("connect",()=>{
+                client2.emit("login", authToken);
+                client2.on("login_success",(status)=>{
+                    expect(status).to.equal(true);
+                    done();
+                });
+            });
+        })
+        .catch( (err) => {
+            done(err);
+        }); 
+    });
+
+    it("should emit messages to server", (done) => {
+        client3 = io(baseUrl, options);
+        client3.emit("login",authToken);
+        client3.on("login_success", (status)=>{
+            expect(status).to.equal(true);
+            client3.emit("panic", { classroom: "asdf", state: true });
```

Comment:
how do you know if this emit works?

Reply:
Do you have a way of testing that an emit was actually emitted? I'm not familiar with socket.io enough to know if there is a way, but I figured if it has gotten to this point without failing out it would have successfully emitted the "panic"